smk/include/versions.smk

```diff
@@ -0,0 +1,493 @@
+
+Authors: Judit Szarvas
+'''
+
+import sys
+from os import path
+
+include: 'config_parser.smk'
+
+def get_smk_filename():
+    if '-s' in sys.argv:
+        i = sys.argv.index('-s')
+    elif '--snakefile' in sys.argv:
+        i = sys.argv.index('--snakefile')
+    smk_name = path.splitext(path.basename(sys.argv[i+1]))[0]
+    return smk_name
+
+def get_version_output(smkfilename):
+    smkfiles = ["QC_0", "QC_1", "QC_2", "assembly", "binning_preparation", "mags_generation", "gene_annotation", "gene_abundance", "data_integration", "module_tester"]
+    if smkfilename in smkfiles:
+        result = "{wd}/output/versions/{smk}.flag".format(wd=working_dir, smk=smkfilename)
+        return(result)
+    else:
+        print('WARNING:', smkfilename, 'not set up for printing software versions')
+    return()
+
+snakefile_name = get_smk_filename()
+
+metaphlan_version = ""
+if "metaphlan_version" in config and config['metaphlan_version'] is not None:
+    metaphlan_version = config['metaphlan_version']
+motus_version = ""
+if "motus_version" in config and config['motus_version'] is not None:
+    motus_version = config['motus_version']
+
+spades_script = 'spades.py' # from conda environment
+if 'METASPADES_custom_build' in config:
+    spades_script = config['METASPADES_custom_build']
+
+ppl_version = ""
+gtdb_version = ""
+if "PHYLOPHLAN_TAXONOMY_VERSION" in config and config["PHYLOPHLAN_TAXONOMY_VERSION"] is not None:
+    ppl_version = config["PHYLOPHLAN_TAXONOMY_VERSION"]
+if "GTDB_TAXONOMY_VERSION" in config and config["GTDB_TAXONOMY_VERSION"] is not None:
+    gtdb_version = config["GTDB_TAXONOMY_VERSION"]
+
+##################################
+#  Rules for TESTING
+##################################
+
+rule test_base:
+    output:
+        temp("{wd}/output/versions/module_tester.flag")
+    resources:
+        mem=1
+    threads:
+        1
+    localrule: True
+    conda:
+        config["minto_dir"]+"/envs/MIntO_base.yml"
+    shell:
+        """
+        echo "MIntO git commit $(cd {minto_dir} && git show --pretty=reference && cd - > /dev/null)" > {working_dir}/output/versions/{snakefile_name}.$(date "+%Y-%m-%d").txt
+        touch {output}
+        """
+
+##################################
+#  Rules for QC_0
+##################################
+
+rule QC_0_base:
+    output:
+        temp("{wd}/output/versions/QC_0_base.flag")
+    resources:
+        mem=1
+    threads:
+        1
+    localrule: True
+    conda:
+        config["minto_dir"]+"/envs/MIntO_base.yml"
+    shell:
+        """
+        echo "MIntO git commit $(cd {minto_dir} && git show --pretty=reference && cd - > /dev/null)" > {working_dir}/output/versions/{snakefile_name}.$(date "+%Y-%m-%d").txt
+        fastqc --version >> {working_dir}/output/versions/{snakefile_name}.$(date "+%Y-%m-%d").txt
+        echo $(fastp --version 2>&1) >> {working_dir}/output/versions/{snakefile_name}.$(date "+%Y-%m-%d").txt
+        touch {output}
```

Comment:
Could we define the filename and use the variable name later? Pathologically, if this runs at 23:59:59, you might make two files in the end.

Reply:
I do that within-rule in a later commit. But not between rules (ie in the same smk but different conda envs), although that could be done too, I think.